src/net.cpp

```diff
@@ -1308,19 +1316,20 @@ void CConnman::SocketHandlerConnected(const std::vector<CNode*>& nodes,
             {
                 bool notify = false;
 
-                // a map to store the bytes by message type
-                mapMsgTypeSize msgtype_bytes;
+                // a map to store each message type mapped to the total message count,
+                // as well as the total number of message bytes
+                std::map<std::string, std::tuple<int, uint64_t>> msgtype_countbytes;
 
                 // initialize all message type values to zero
                 for (const std::string& msg : getAllNetMessageTypes())
-                    msgtype_bytes[msg] = 0;
-                msgtype_bytes[NET_MESSAGE_TYPE_OTHER] = 0;
+                    msgtype_countbytes[msg] = std::make_tuple(0, 0);
+                msgtype_countbytes[NET_MESSAGE_TYPE_OTHER] = std::make_tuple(0, 0);
 
-                if (!pnode->ReceiveMsgBytes({pchBuf, (size_t)nBytes}, notify, msgtype_bytes)) {
+                if (!pnode->ReceiveMsgBytes({pchBuf, (size_t)nBytes}, notify, msgtype_countbytes)) {
                     pnode->CloseSocketDisconnect();
                 }
                 RecordBytesRecv(nBytes);
-                RecordBytesRecvByMsgType(msgtype_bytes);
+                RecordMsgStatsRecv(msgtype_countbytes, pnode->m_conn_type, pnode->ConnectedThroughNetwork());
```

Comment:
Good catch! Addressed in https://github.com/satsie/bitcoin/pull/3/commits/4c3b8216e90596036beda4e1f03d65afd96a3499

To take it a step further, I took a peek at the counterpart method, `RecordBytesSent()` and saw it:
- uses the `m_total_bytes_sent_mutex`, and 
- checks the current time to see if it needs to reset something called a max outbound cycle (`nMaxOutboundCycleStartTime` and `nMaxOutboundTotalBytesSentInCycle`)

After seeing that, I figured it was best to leave `RecordBytesSent()` alone. However, I can see an argument to make `RecordBytesSent()` private, and call it from within `RecordMsgStatsSent()`. Any thoughts on this? Would you consider such a change out of scope?

Reply:
**Update**: I unfortunately have to undo that change to move `RecordBytesRecv()` into `RecordMsgStatsRecv()`. I was working on [this comment](https://github.com/satsie/bitcoin/pull/3#discussion_r1069234557) to remove the output parameter from `ReceiveMsgBytes()` and ended up having to push the call to record the number of bytes, `RecordMsgStatsRecv()`. 

I am now waiting for the message to be "published" to `CNode`'s `vRecvMsg` list before invoking `RecordMsgStatsRecv()`. `vRecvMsg` only contains completed CNetMessages.

Upon doing this I noticed `test_buffer()` in `p2p_invalid_messsages.py` failing. This is because that test sends a message in two pieces, and `RecordMsgStatsRecv()` doesn't get called until the full message has made it through. Therefore, the `getnettotals`.`totalbytesrecv` value that the test is looking at, is not updated until the second half of the message is sent.

Seems like the current behavior is `getnettotals`.`totalbytesrecv` is accurate even when messages are split into pieces, but my new `getnetmsgstats` needs to wait for messages to fully come through before updating. This makes sense to me because I'm collecting message counts, which are whole numbers